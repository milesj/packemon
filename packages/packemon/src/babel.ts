import fs from 'fs-extra';
import { TransformOptions as ConfigStructure } from '@babel/core';
import { Path, toArray } from '@boost/common';
import { FeatureFlags } from './types';
import {
	CodeArtifact,
	DEFAULT_SUPPORT,
	Format,
	getBabelInputConfig,
	getBabelOutputConfig,
	Package,
	PackemonPackage,
	Platform,
	Project,
	Support,
} from '.';

const format = (process.env.PACKEMON_FORMAT ?? 'lib') as Format;
const support = (process.env.PACKEMON_SUPPORT ?? DEFAULT_SUPPORT) as Support;
const project = new Project(process.cwd());

function getBabelConfig(artifact: CodeArtifact, featureFlags: FeatureFlags): ConfigStructure {
	const inputConfig = getBabelInputConfig(artifact, featureFlags);
	const outputConfig = getBabelOutputConfig(
		artifact.platform,
		artifact.support,
		artifact.builds[0].format,
		featureFlags,
	);

	return {
		// Input must come first
		plugins: [...inputConfig.plugins!, ...outputConfig.plugins!],
		// Env must come first
		presets: [...outputConfig.presets!, ...inputConfig.presets!],
	};
}

export interface ConfigOptions {
	format?: Format;
	platform?: Platform;
	support?: Support;
}

export function createConfig(folder: string, options: ConfigOptions = {}): ConfigStructure {
	const path = new Path(folder);
	const contents = fs.readJsonSync(path.append('package.json').path()) as PackemonPackage;

	// Create package and configs
	const pkg = new Package(project, path, contents);

	if (pkg.packageJson.packemon) {
		pkg.setConfigs(toArray(pkg.packageJson.packemon));
	}

	// Determine the lowest platform to support
	const platforms = pkg.configs.map((config) => config.platform);
	let lowestPlatform: Platform = 'node';

	// istanbul ignore next
	if (platforms.includes('browser')) {
		lowestPlatform = 'browser';
	} else if (platforms.includes('native')) {
		lowestPlatform = 'native';
	}

	// Generate artifact and builds
	const artifact = new CodeArtifact(pkg, [{ format: options.format ?? format }]);
	artifact.bundle = false;
	artifact.platform = options.platform ?? lowestPlatform;
	artifact.support = options.support ?? support;

	return getBabelConfig(artifact, pkg.getFeatureFlags());
}

export function createRootConfig(options?: ConfigOptions): ConfigStructure {
	const config = createConfig(process.cwd(), options);

	return {
		...config,
		babelrc: true,
		babelrcRoots: project.getWorkspaceGlobs({ relative: true }),
		// Support React Native libraries by default
		overrides: [
			{
				presets: ['@babel/preset-flow'],
				test: /node_modules\/((jest-)?react-native|@react-native(-community)?)/iu,
			},
		],
	};
}
