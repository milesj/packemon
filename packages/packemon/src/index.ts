/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './Artifact';
export * from './babel/config';
export * from './CodeArtifact';
export * from './commands/Base';
export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Init';
export * from './commands/Pack';
export * from './commands/Scaffold';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './Package';
export * from './PackageValidator';
export * from './Packemon';
export * from './Project';
export * from './rollup/config';
export * from './types';
export * from './TypesArtifact';
