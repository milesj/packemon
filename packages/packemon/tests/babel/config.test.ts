/* eslint-disable no-param-reassign */

// import { createConfig, createRootConfig } from '../../src/babel';
import { getBabelInputConfig, getBabelOutputConfig } from '../../src/babel/config';
import { Format, Platform, Support } from '../../src/types';

jest.mock('../../src/babel/resolve', () => ({
	resolve: (name: string) => name,
	resolveFromBabel: (name: string) => name,
}));

const SUPPORTS: Support[] = ['legacy', 'stable', 'current', 'experimental'];

describe('getBabelInputConfig()', () => {
	const bundleArtifact: any = {
		features: {},
		package: { hasDependency: () => false },
	};

	it('includes no plugins or presets by default', () => {
		expect(getBabelInputConfig(bundleArtifact, {})).toMatchSnapshot();
	});

	it('includes react preset if `react` feature flag is true', () => {
		expect(getBabelInputConfig(bundleArtifact, { react: 'classic' }).presets).toMatchSnapshot();
		expect(getBabelInputConfig(bundleArtifact, { react: 'automatic' }).presets).toMatchSnapshot();
	});

	it('includes solid preset if `solid` feature flag is true', () => {
		expect(getBabelInputConfig(bundleArtifact, { solid: true }).presets).toMatchSnapshot();
	});

	it('includes flow preset if `flow` feature flag is true', () => {
		expect(getBabelInputConfig(bundleArtifact, { flow: true }).presets).toMatchSnapshot();
	});

	it('includes typescript preset if `typescript` feature flag is true', () => {
		expect(getBabelInputConfig(bundleArtifact, { typescript: true }).presets).toMatchSnapshot();
	});

	it('includes typescript decorators if `typescript` and `decorators` feature flag is true', () => {
		expect(
			getBabelInputConfig(bundleArtifact, { decorators: true, typescript: true }),
		).toMatchSnapshot();
	});

	it('doesnt include typescript decorators if `typescript` feature flag is false', () => {
		expect(
			getBabelInputConfig(bundleArtifact, { decorators: true, typescript: false }),
		).toMatchSnapshot();
	});

	it('supports private properties with decorators if dep exists', () => {
		const spy = jest.spyOn(bundleArtifact.package, 'hasDependency').mockImplementation(() => true);

		expect(
			getBabelInputConfig(bundleArtifact, { decorators: true, typescript: true }),
		).toMatchSnapshot();

		spy.mockRestore();
	});

	it('can mutate config', () => {
		expect(
			getBabelInputConfig(
				bundleArtifact,
				{},
				{
					babelInput(config) {
						config.rootMode = 'upward';
						config.plugins?.push('babel-plugin-example');
					},
					babelOutput(config) {
						config.plugins?.push('babel-plugin-shouldnt-show-up');
					},
				},
			),
		).toMatchSnapshot();
	});
});

function renderPresetEnv(platform: Platform, format: Format, support: Support) {
	// eslint-disable-next-line jest/require-top-level-describe
	test(`handles preset-env: ${platform} + ${format} + ${support}`, () => {
		expect(getBabelOutputConfig(platform, support, format, {})?.presets?.[0]).toMatchSnapshot();
	});
}

describe('getBabelOutputConfig()', () => {
	SUPPORTS.forEach((support) => {
		renderPresetEnv('native', 'lib', support);

		(['lib', 'cjs', 'mjs'] as const).forEach((format) => {
			renderPresetEnv('node', format, support);
		});

		(['lib', 'esm', 'umd'] as const).forEach((format) => {
			renderPresetEnv('browser', format, support);
		});
	});

	it('errors for invalid platform', () => {
		expect(() =>
			// @ts-expect-error Unknown platform
			getBabelOutputConfig('unknown', 'stable', 'lib', {}),
		).toThrow('Unknown platform "unknown".');
	});

	it('transforms async/await to promises when `browser` or `native`', () => {
		expect(getBabelOutputConfig('browser', 'stable', 'lib', {})).toMatchSnapshot();

		expect(getBabelOutputConfig('native', 'experimental', 'lib', {})).toMatchSnapshot();
	});

	it('uses built-in destructuring and object spread when `current` or `experimental`', () => {
		expect(getBabelOutputConfig('node', 'current', 'lib', {})).toMatchSnapshot();

		expect(getBabelOutputConfig('node', 'experimental', 'lib', {})).toMatchSnapshot();
	});

	it('sets `parserOpts.strictMode` based on `strict` feature flag', () => {
		expect(getBabelOutputConfig('node', 'stable', 'lib', { strict: true })).toEqual(
			expect.objectContaining({
				parserOpts: {
					sourceType: 'unambiguous',
					strictMode: true,
				},
			}),
		);
	});

	it('can mutate config', () => {
		expect(
			getBabelOutputConfig(
				'browser',
				'stable',
				'lib',
				{},
				{
					babelInput(config) {
						config.plugins?.push('babel-plugin-shouldnt-show-up');
					},
					babelOutput(config) {
						config.rootMode = 'upward';
						config.plugins?.push('babel-plugin-example');
					},
				},
			),
		).toMatchSnapshot();
	});

	it('passes build params to config', () => {
		const spy = jest.fn();

		getBabelOutputConfig(
			'browser',
			'stable',
			'lib',
			{},
			{
				babelOutput: spy,
			},
		);

		expect(spy).toHaveBeenCalledWith(expect.any(Object), {
			features: {},
			format: 'lib',
			platform: 'browser',
			support: 'stable',
		});
	});
});

// describe('createRootConfig()', () => {
// 	it('returns the correct config', () => {
// 		expect(createRootConfig()).toMatchSnapshot();
// 	});
// });

// describe('createConfig()', () => {
// 	it('returns the correct config', () => {
// 		expect(createConfig(process.cwd())).toMatchSnapshot();
// 	});
// });
