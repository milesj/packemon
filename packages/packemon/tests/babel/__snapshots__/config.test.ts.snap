// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig() returns the correct config 1`] = `
Object {
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-class-properties",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-methods",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-property-in-object",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "12.17.0",
        },
        "useBuiltIns": false,
      },
    ],
    Array [
      "@babel/preset-typescript",
      Object {
        "allowDeclareFields": true,
      },
    ],
  ],
}
`;

exports[`createRootConfig() returns the correct config 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": Array [
    "packages/*",
    "website",
  ],
  "overrides": Array [
    Object {
      "presets": Array [
        "@babel/preset-flow",
      ],
      "test": /node_modules\\\\/\\(\\(jest-\\)\\?react-native\\|@react-native\\(-community\\)\\?\\)/iu,
    },
  ],
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-class-properties",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-methods",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-property-in-object",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "12.17.0",
        },
        "useBuiltIns": false,
      },
    ],
    Array [
      "@babel/preset-typescript",
      Object {
        "allowDeclareFields": true,
      },
    ],
  ],
}
`;

exports[`getBabelInputConfig() doesnt include typescript decorators if \`typescript\` feature flag is false 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [],
  "presets": Array [],
}
`;

exports[`getBabelInputConfig() includes flow preset if \`flow\` feature flag is true 1`] = `
Array [
  Array [
    "@babel/preset-flow",
    Object {
      "allowDeclareFields": true,
    },
  ],
]
`;

exports[`getBabelInputConfig() includes no plugins or presets by default 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [],
  "presets": Array [],
}
`;

exports[`getBabelInputConfig() includes react preset if \`react\` feature flag is true 1`] = `
Array [
  Array [
    "@babel/preset-react",
    Object {
      "development": true,
      "throwIfNamespace": true,
    },
  ],
]
`;

exports[`getBabelInputConfig() includes typescript decorators if \`typescript\` and \`decorators\` feature flag is true 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-class-properties",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-methods",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-property-in-object",
      Object {
        "loose": true,
      },
    ],
  ],
  "presets": Array [
    Array [
      "@babel/preset-typescript",
      Object {
        "allowDeclareFields": true,
      },
    ],
  ],
}
`;

exports[`getBabelInputConfig() includes typescript preset if \`typescript\` feature flag is true 1`] = `
Array [
  Array [
    "@babel/preset-typescript",
    Object {
      "allowDeclareFields": true,
    },
  ],
]
`;

exports[`getBabelInputConfig() supports private properties with decorators if dep exists 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-proposal-decorators",
      Object {
        "legacy": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-class-properties",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-methods",
      Object {
        "loose": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-private-property-in-object",
      Object {
        "loose": true,
      },
    ],
  ],
  "presets": Array [
    Array [
      "@babel/preset-typescript",
      Object {
        "allowDeclareFields": true,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "> 1%",
        "not dead",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "defaults",
        "not IE 11",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "> 1%",
        "not dead",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "defaults",
        "not IE 11",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "> 1%",
        "not dead",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "defaults",
        "not IE 11",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 12",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 14",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 8",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 10",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "14.16.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "16.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.17.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "14.16.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "16.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.17.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "14.16.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "16.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.17.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    "@babel/plugin-proposal-logical-assignment-operators",
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "@babel/plugin-proposal-optional-chaining",
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "browsers": Array [
            "defaults",
            "not IE 11",
          ],
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 2`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    "@babel/plugin-proposal-logical-assignment-operators",
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "@babel/plugin-proposal-optional-chaining",
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "browsers": "iOS 14",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "14.16.0",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 2`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": true,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    "babel-plugin-env-constants",
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "16.0.0",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;
